File: src/agent/read-policy.ts
import { lstat, open, realpath } from "node:fs/promises"
import { isAbsolute, resolve } from "node:path"
import { isWithin } from "../config.ts"

export const MAX_READ_PATH_BYTES = 4096
export const MAX_READ_FILE_BYTES = 1_000_000
export const MAX_READ_LINES = 2000
export const MAX_READ_LINE_NUMBER = 1_000_000

export interface ReadInput {
  path: string
  offset?: number
  limit?: number
}

export interface ReadResult {
  path: string
  size: number
  totalLines: number
  startLine: number
  endLine: number
  truncated: boolean
  text: string
}

interface AuthorizedRead {
  path: string
  size: number
  offset: number
  limit: number | undefined
}

export class ReadPolicy {
  constructor(
    private readonly roots: string[],
    readonly maxOutputBytes: number,
  ) {}

  async authorize(input: ReadInput, cwd: string): Promise<AuthorizedRead> {
    validateInput(input)
    const requested = isAbsolute(input.path)
      ? resolve(input.path)
      : resolve(cwd, input.path)
    if (!isWithin(requested, this.roots))
      throw new Error(`file is outside approved roots: ${input.path}`)

    let canonical: string
    try {
      canonical = await realpath(requested)
    } catch {
      throw new Error(`file is unavailable: ${input.path}`)
    }
    if (!isWithin(canonical, this.roots))
      throw new Error(
        `canonical file path is outside approved roots: ${input.path}`,
      )

    const stat = await lstat(canonical)
    if (!stat.isFile())
      throw new Error(`path is not a regular file: ${input.path}`)
    if (stat.size > MAX_READ_FILE_BYTES) {
      throw new Error(
        `file exceeds the ${MAX_READ_FILE_BYTES} byte read limit: ${input.path}`,
      )
    }

    return {
      path: canonical,
      size: stat.size,
      offset: input.offset ?? 1,
      limit: input.limit,
    }
  }

  async read(input: ReadInput, cwd: string): Promise<ReadResult> {
    const authorized = await this.authorize(input, cwd)
    const bytes = await readBoundedFile(authorized.path)

    let text: string
    try {
      text = new TextDecoder("utf-8", { fatal: true }).decode(bytes)
    } catch {
      throw new Error(`file is not valid UTF-8 text: ${input.path}`)
    }
    const lines = text.split(/\r?\n/)
    if (authorized.offset > lines.length) {
      throw new Error(
        `offset ${authorized.offset} is beyond end of file (${lines.length} lines)`,
      )
    }

    const startIndex = authorized.offset - 1
    const endIndex = Math.min(
      lines.length,
      startIndex + (authorized.limit ?? lines.length),
    )
    const formatted = formatLines(
      lines,
      startIndex,
      endIndex,
      this.maxOutputBytes,
    )

    return {
      path: authorized.path,
      size: bytes.byteLength,
      totalLines: lines.length,
      startLine: authorized.offset,
      endLine: formatted.endLine,
      truncated: formatted.truncated || endIndex < lines.length,
      text: formatted.text,
    }
  }
}

async function readBoundedFile(path: string): Promise<Buffer> {
  const file = await open(path, "r")
  try {
    const stat = await file.stat()
    if (!stat.isFile()) throw new Error(`path is not a regular file: ${path}`)
    if (stat.size > MAX_READ_FILE_BYTES) {
      throw new Error(
        `file exceeds the ${MAX_READ_FILE_BYTES} byte read limit: ${path}`,
      )
    }

    const buffer = Buffer.alloc(MAX_READ_FILE_BYTES + 1)
    let size = 0
    while (size < buffer.byteLength) {
      const { bytesRead } = await file.read(
        buffer,
        size,
        buffer.byteLength - size,
        size,
      )
      if (bytesRead === 0) break
      size += bytesRead
    }
    if (size > MAX_READ_FILE_BYTES) {
      throw new Error(
        `file exceeds the ${MAX_READ_FILE_BYTES} byte read limit: ${path}`,
      )
    }
    return buffer.subarray(0, size)
  } finally {
    await file.close()
  }
}

function validateInput(input: ReadInput): void {
  if (
    input.path.length === 0 ||
    Buffer.byteLength(input.path) > MAX_READ_PATH_BYTES
  ) {
    throw new Error("file path length is outside read policy bounds")
  }
  if (input.path.includes("\0")) throw new Error("NUL bytes are forbidden")
  if (input.offset !== undefined) {
    if (
      !Number.isInteger(input.offset) ||
      input.offset < 1 ||
      input.offset > MAX_READ_LINE_NUMBER
    ) {
      throw new Error("offset is outside read policy bounds")
    }
  }
  if (input.limit !== undefined) {
    if (
      !Number.isInteger(input.limit) ||
      input.limit < 1 ||
      input.limit > MAX_READ_LINES
    ) {
      throw new Error("limit is outside read policy bounds")
    }
  }
}

function formatLines(
  lines: string[],
  startIndex: number,
  endIndex: number,
  maxBytes: number,
): { text: string; endLine: number; truncated: boolean } {
  const output: string[] = []
  for (let index = startIndex; index < endIndex; index += 1) {
    const lineNumber = index + 1
    const rendered = `${lineNumber}\t${lines[index]}`
    const hasMore = index + 1 < lines.length
    const continuation = hasMore
      ? `\n[Output truncated at ${maxBytes} bytes. Showing lines ${startIndex + 1}-${lineNumber} of ${lines.length}. Use offset=${lineNumber + 1} to continue.]`
      : ""
    const candidate = [...output, rendered].join("\n") + continuation
    if (Buffer.byteLength(candidate) <= maxBytes) {
      output.push(rendered)
      continue
    }

    const priorEnd = lineNumber - 1
    if (output.length > 0) {
      const notice = `[Output truncated at ${maxBytes} bytes. Showing lines ${startIndex + 1}-${priorEnd} of ${lines.length}. Use offset=${lineNumber} to continue.]`
      return {
        text: `${output.join("\n")}\n${notice}`,
        endLine: priorEnd,
        truncated: true,
      }
    }

    const notice = `\n[Output truncated at ${maxBytes} bytes on line ${lineNumber} of ${lines.length}. Use offset=${lineNumber} to reread the line.]`
    const prefix = `${lineNumber}\t`
    const available = Math.max(
      0,
      maxBytes - Buffer.byteLength(prefix) - Buffer.byteLength(notice),
    )
    return {
      text: `${prefix}${truncateUtf8(lines[index] ?? "", available)}${notice}`,
      endLine: lineNumber,
      truncated: true,
    }
  }

  const endLine = Math.max(startIndex + 1, endIndex)
  if (endIndex < lines.length) {
    output.push(
      `[Showing lines ${startIndex + 1}-${endLine} of ${lines.length}. Use offset=${endLine + 1} to continue.]`,
    )
  }
  return { text: output.join("\n"), endLine, truncated: false }
}

function truncateUtf8(value: string, maxBytes: number): string {
  const bytes = Buffer.from(value)
  if (bytes.byteLength <= maxBytes) return value
  let end = maxBytes
  while (end > 0) {
    try {
      return new TextDecoder("utf-8", { fatal: true }).decode(
        bytes.subarray(0, end),
      )
    } catch {
      end -= 1
    }
  }
  return ""
}

File: src/agent/read_policy.rs
use std::ffi::CString;
use std::fs::File;
use std::io::Read;
use std::os::fd::{FromRawFd, OwnedFd};
use std::os::unix::ffi::OsStrExt;
use std::path::{Component, Path, PathBuf};

use anyhow::{Context, Result, bail};

use crate::config::is_within;

pub const MAX_READ_PATH_BYTES: usize = 4096;
pub const MAX_READ_FILE_BYTES: usize = 1_000_000;
pub const MAX_READ_LINES: usize = 2000;
pub const MAX_READ_LINE_NUMBER: usize = 1_000_000;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ReadInput {
    pub path: String,
    pub offset: Option<usize>,
    pub limit: Option<usize>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ReadResult {
    pub path: PathBuf,
    pub size: usize,
    pub total_lines: usize,
    pub start_line: usize,
    pub end_line: usize,
    pub truncated: bool,
    pub text: String,
}

#[derive(Debug)]
pub struct AuthorizedRead {
    path: PathBuf,
    file: File,
    size: usize,
    offset: usize,
    limit: Option<usize>,
}

#[derive(Clone, Debug)]
pub struct ReadPolicy {
    roots: Vec<PathBuf>,
    pub max_output_bytes: usize,
}

impl ReadPolicy {
    pub fn new(roots: Vec<PathBuf>, max_output_bytes: usize) -> Result<Self> {
        let mut approved = Vec::new();
        for root in roots {
            let lexical = normalize(&root)?;
            let canonical = std::fs::canonicalize(&root).with_context(|| {
                format!("approved read root is unavailable: {}", root.display())
            })?;
            approved.push(lexical);
            approved.push(canonical);
        }
        approved.sort();
        approved.dedup();
        Ok(Self {
            roots: approved,
            max_output_bytes,
        })
    }

    pub fn authorize(&self, input: &ReadInput, cwd: &Path) -> Result<AuthorizedRead> {
        validate_input(input)?;
        let requested = if Path::new(&input.path).is_absolute() {
            normalize(Path::new(&input.path))?
        } else {
            normalize(&cwd.join(&input.path))?
        };
        if !is_within(&requested, &self.roots) {
            bail!("file is outside approved roots: {}", input.path);
        }
        let canonical = std::fs::canonicalize(&requested)
            .map_err(|_| anyhow::anyhow!("file is unavailable: {}", input.path))?;
        if !is_within(&canonical, &self.roots) {
            bail!(
                "canonical file path is outside approved roots: {}",
                input.path
            );
        }

        let file = open_absolute_no_follow(&canonical, false)
            .map_err(|_| anyhow::anyhow!("file is unavailable: {}", input.path))?;
        let metadata = file
            .metadata()
            .with_context(|| format!("inspect opened file: {}", input.path))?;
        if !metadata.is_file() {
            bail!("path is not a regular file: {}", input.path);
        }
        let size = usize::try_from(metadata.len()).unwrap_or(usize::MAX);
        if size > MAX_READ_FILE_BYTES {
            bail!(
                "file exceeds the {MAX_READ_FILE_BYTES} byte read limit: {}",
                input.path
            );
        }
        Ok(AuthorizedRead {
            path: canonical,
            file,
            size,
            offset: input.offset.unwrap_or(1),
            limit: input.limit,
        })
    }

    pub fn read(&self, input: &ReadInput, cwd: &Path) -> Result<ReadResult> {
        let authorized = self.authorize(input, cwd)?;
        let AuthorizedRead {
            path,
            file,
            size: authorized_size,
            offset,
            limit,
        } = authorized;
        let mut bytes = Vec::with_capacity(authorized_size.saturating_add(1));
        file.take((MAX_READ_FILE_BYTES + 1) as u64)
            .read_to_end(&mut bytes)
            .with_context(|| format!("read file: {}", input.path))?;
        if bytes.len() > MAX_READ_FILE_BYTES {
            bail!(
                "file exceeds the {MAX_READ_FILE_BYTES} byte read limit: {}",
                input.path
            );
        }
        let text = String::from_utf8(bytes)
            .map_err(|_| anyhow::anyhow!("file is not valid UTF-8 text: {}", input.path))?;
        let lines = text
            .split('\n')
            .map(|line| line.strip_suffix('\r').unwrap_or(line))
            .collect::<Vec<_>>();
        if offset > lines.len() {
            bail!(
                "offset {offset} is beyond end of file ({} lines)",
                lines.len()
            );
        }
        let start = offset - 1;
        let end = lines
            .len()
            .min(start.saturating_add(limit.unwrap_or(lines.len())));
        let formatted = format_lines(&lines, start, end, self.max_output_bytes);
        Ok(ReadResult {
            path,
            size: text.len(),
            total_lines: lines.len(),
            start_line: offset,
            end_line: formatted.end_line,
            truncated: formatted.truncated || end < lines.len(),
            text: formatted.text,
        })
    }
}

fn validate_input(input: &ReadInput) -> Result<()> {
    if input.path.is_empty() || input.path.len() > MAX_READ_PATH_BYTES {
        bail!("file path length is outside read policy bounds");
    }
    if input.path.as_bytes().contains(&0) {
        bail!("NUL bytes are forbidden");
    }
    if input
        .offset
        .is_some_and(|value| value == 0 || value > MAX_READ_LINE_NUMBER)
    {
        bail!("offset is outside read policy bounds");
    }
    if input
        .limit
        .is_some_and(|value| value == 0 || value > MAX_READ_LINES)
    {
        bail!("limit is outside read policy bounds");
    }
    Ok(())
}

struct FormattedLines {
    text: String,
    end_line: usize,
    truncated: bool,
}

fn format_lines(lines: &[&str], start: usize, end: usize, max_bytes: usize) -> FormattedLines {
    let mut output = Vec::new();
    for index in start..end {
        let line_number = index + 1;
        let rendered = format!("{line_number}\t{}", lines[index]);
        let has_more = index + 1 < lines.len();
        let continuation = if has_more {
            format!(
                "\n[Output truncated at {max_bytes} bytes. Showing lines {}-{line_number} of {}. Use offset={} to continue.]",
                start + 1,
                lines.len(),
                line_number + 1
            )
        } else {
            String::new()
        };
        let candidate = format!(
            "{}{}{}",
            output.join("\n"),
            if output.is_empty() { "" } else { "\n" },
            rendered
        );
        if candidate.len() + continuation.len() <= max_bytes {
            output.push(rendered);
            continue;
        }
        let prior_end = line_number - 1;
        if !output.is_empty() {
            let notice = format!(
                "[Output truncated at {max_bytes} bytes. Showing lines {}-{prior_end} of {}. Use offset={line_number} to continue.]",
                start + 1,
                lines.len()
            );
            return FormattedLines {
                text: format!("{}\n{notice}", output.join("\n")),
                end_line: prior_end,
                truncated: true,
            };
        }
        let notice = format!(
            "\n[Output truncated at {max_bytes} bytes on line {line_number} of {}. Use offset={line_number} to reread the line.]",
            lines.len()
        );
        let prefix = format!("{line_number}\t");
        let available = max_bytes.saturating_sub(prefix.len() + notice.len());
        return FormattedLines {
            text: format!("{prefix}{}{notice}", truncate_utf8(lines[index], available)),
            end_line: line_number,
            truncated: true,
        };
    }

    let end_line = (start + 1).max(end);
    if end < lines.len() {
        output.push(format!(
            "[Showing lines {}-{end_line} of {}. Use offset={} to continue.]",
            start + 1,
            lines.len(),
            end_line + 1
        ));
    }
    FormattedLines {
        text: output.join("\n"),
        end_line,
        truncated: false,
    }
}

fn truncate_utf8(value: &str, max_bytes: usize) -> &str {
    if value.len() <= max_bytes {
        return value;
    }
    let mut end = max_bytes.min(value.len());
    while !value.is_char_boundary(end) {
        end -= 1;
    }
    &value[..end]
}

fn normalize(path: &Path) -> Result<PathBuf> {
    let absolute = if path.is_absolute() {
        path.to_path_buf()
    } else {
        std::env::current_dir()?.join(path)
    };
    let mut normalized = PathBuf::new();
    for component in absolute.components() {
        match component {
            Component::CurDir => {}
            Component::ParentDir => {
                normalized.pop();
            }
            other => normalized.push(other.as_os_str()),
        }
    }
    Ok(normalized)
}

pub(crate) fn open_absolute_no_follow(path: &Path, directory: bool) -> std::io::Result<File> {
    if !path.is_absolute() {
        return Err(std::io::Error::new(
            std::io::ErrorKind::InvalidInput,
            "path must be absolute",
        ));
    }
    let root = CString::new("/").expect("root path has no NUL");
    let root_fd = unsafe {
        libc::open(
            root.as_ptr(),
            libc::O_RDONLY | libc::O_DIRECTORY | libc::O_CLOEXEC,
        )
    };
    if root_fd < 0 {
        return Err(std::io::Error::last_os_error());
    }
    let mut current = unsafe { OwnedFd::from_raw_fd(root_fd) };
    let components = path
        .components()
        .filter_map(|component| match component {
            Component::Normal(value) => Some(value),
            _ => None,
        })
        .collect::<Vec<_>>();
    if components.is_empty() {
        return Ok(File::from(current));
    }
    for (index, component) in components.iter().enumerate() {
        let name = CString::new(component.as_bytes()).map_err(|_| {
            std::io::Error::new(std::io::ErrorKind::InvalidInput, "path contains NUL")
        })?;
        let last = index + 1 == components.len();
        let mut flags = libc::O_RDONLY | libc::O_CLOEXEC | libc::O_NOFOLLOW;
        if !last || directory {
            flags |= libc::O_DIRECTORY;
        }
        let fd = unsafe {
            libc::openat(
                std::os::fd::AsRawFd::as_raw_fd(&current),
                name.as_ptr(),
                flags,
            )
        };
        if fd < 0 {
            return Err(std::io::Error::last_os_error());
        }
        current = unsafe { OwnedFd::from_raw_fd(fd) };
    }
    Ok(File::from(current))
}
